File: ports/js/src/capabilities.js
// Capability model (spec PROTOCOL.md sec 8).
//
// Direct port of packages/confine/capabilities.fard's operation_allowed().
// No default-allow: an actor absent from the map, or an operation absent
// from that actor's explicit list, returns false.

import { taggedDigest } from "./hash.js";

export function operationAllowed(capabilities, actorId, operation) {
  const actor = capabilities.actors[actorId];
  if (!actor) return false;
  return actor.operations.includes(operation);
}

export function actorRole(capabilities, actorId) {
  const actor = capabilities.actors[actorId];
  return actor ? actor.role : null;
}


export function capabilitiesToValue(capabilities) {
  const actors = {};
  for (const [k, v] of Object.entries(capabilities.actors)) {
    actors[k] = { role: v.role, operations: [...v.operations] };
  }
  const operations = {};
  for (const [k, v] of Object.entries(capabilities.operationsRegistry ?? {})) {
    operations[k] = v;
  }
  return { actors, operations };
}

export function capabilitiesDigest(capabilities) {
  return taggedDigest("confine.capabilities.v1", capabilitiesToValue(capabilities));
}

File: ports/python/confine_core/capabilities.py
"""Capability model (spec PROTOCOL.md sec 8).

Direct port of packages/confine/capabilities.fard's operation_allowed().
No default-allow: an actor absent from the map, or an operation absent
from that actor's explicit list, returns False.
"""

from dataclasses import dataclass, field


@dataclass
class Actor:
    role: str
    operations: list


@dataclass
class Capabilities:
    actors: dict
    operations_registry: dict = None  # populated separately; see digest()

    def __post_init__(self):
        if self.operations_registry is None:
            self.operations_registry = {}

    def to_dict(self):
        return {
            "actors": {k: {"role": v.role, "operations": list(v.operations)} for k, v in self.actors.items()},
            "operations": dict(sorted(self.operations_registry.items())),
        }


def operation_allowed(capabilities: Capabilities, actor_id: str, operation: str) -> bool:
    actor = capabilities.actors.get(actor_id)
    if actor is None:
        return False
    return operation in actor.operations


def actor_role(capabilities: Capabilities, actor_id: str):
    actor = capabilities.actors.get(actor_id)
    return actor.role if actor else None


def digest(capabilities: Capabilities) -> str:
    from .hash import tagged_digest
    return tagged_digest("confine.capabilities.v1", capabilities.to_dict())
